nexus/src/db/datastore.rs

```diff
@@ -143,6 +142,53 @@ impl DataStore {
         Ok(self.pool.pool())
     }
 
+    /// Stores a new rack in the database.
+    ///
+    /// This function is a no-op if the rack already exists.
+    pub async fn rack_insert(
+        &self,
+        opctx: &OpContext,
+        rack: &Rack,
+    ) -> Result<Rack, Error> {
+        use db::schema::rack::dsl;
+
+        diesel::insert_into(dsl::rack)
+            .values(rack.clone())
+            .on_conflict(dsl::id)
+            .do_update()
+            // This is a no-op, since we conflicted on the ID.
+            .set(dsl::id.eq(excluded(dsl::id)))
```

Comment:
https://github.com/oxidecomputer/omicron/issues/1168 for an explanation of why I'm not calling `.do_nothing()` here. Doing so would cause the test I added to fail.